### PR TITLE
Create box tests

### PR DESCRIPTION
**Pull request by billyhao12**
I am writing tests for the CreateBox component located in react/src/views/CreateBox.tsx.
A previous error was resolved by using a MockedProvider provided by Apollo.
Now, the application is getting a different error located in the CreateBox component.

The error says "TypeError: Cannot read property 'location' of undefined" and points to line 18 in CreateBox.tsx:
```
const location: LocationState = useLocation();
```

Not sure how to resolve this. Any help is much appreciated.
The issue at this link looks similar: https://github.com/testing-library/react-testing-library/issues/493

### CODE REVIEW

**Comment by billyhao12** [comment]
This issue has been resolved with help from Hans and this resource: https://testing-library.com/docs/example-react-router. The react-router-dom history is now being mocked with createMemoryHistory, imported from "history".

**Comment by billyhao12** [comment]
I am getting a new error related to the qrUrl. After running the test, it says "TypeError: Cannot read property 'split' of undefined" and points to line 20 of CreateBox.tsx. This means that qrUrl is not defined for some reason. Any idea what I'm missing here?

test-error-2

**Comment by billyhao12** [comment]
The previous issue with the qrUrl has been resolved by pushing a URL to the history constant using `createMemoryHistory()`. I also pushed an example qr code into the history constant. Another "MutationObserver is not defined" error was resolved by installing mutationobserver-shim into the dev dependencies and importing it into CreateBox.test.js.

**Comment by billyhao12** [comment]
New error:
I am trying to test that the form element exists in CreateBox.tsx by targeting it with a test id. However, the test file is unable to find the element.

testId-error

**Comment by aerinsol** [comment]
Nice work @billyhao12 !

**Comment by billyhao12** [comment]
In my last commit, I am trying to test if the required form fields are prohibiting submission by looking for the form element in the DOM. If the form element is present, then the form has not been submitted. If the form element is no longer present, then the form has been submitted.

This method of testing does not appear to be working since the test expecting the form to no longer exist after a successful submission is still detecting the presence of a form.

File: react/src/views/CreateBox/CreateBox.test.js [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+];
+
+describe("Renders components", () => {
+  let component = null;
+  beforeEach(() => {
+    const history = createMemoryHistory();
+    const state = { qr: "barcode=387b0f0f5e62cebcafd48383035a92a" };
+    history.push("/create-box", state);
+
+    component = render(
```
Comment (HaGuesto) [inline react/src/views/CreateBox/CreateBox.test.js:45]: I would strongly suggest that you create a `test-utils` file where you define this custom renderer as a function with the variables (mocks and history)

Every test will most likely need a similar renderer and you already duplicate the definition of it 5 times in this file.

Here, a few steps how to do that
https://testing-library.com/docs/react-testing-library/setup#custom-render

File: react/src/views/CreateBox/CreateBox.test.js [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+  });
+
+  it("renders a submit button titled, `do the mutation`", () => {
+    expect(component.getByRole("button", { name: /do the mutation/i })).toHaveTextContent(
+      "do the mutation",
+    );
+  });
+});
+
+describe("Submission result screen is shown the right way", () => {
```
Comment (HaGuesto) [inline react/src/views/CreateBox/CreateBox.test.js:81]: Created box is displayed correctly

File: react/src/views/CreateBox/CreateBox.test.js [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+          box_id: 456,
+          product_id: 123,
+          items: 50,
+        },
+      },
+    },
+  },
+];
+
+describe("Renders components", () => {
```
Comment (HaGuesto) [inline react/src/views/CreateBox/CreateBox.test.js:38]: Renders CreateBox component correctly

File: react/src/views/CreateBox/CreateBox.test.js [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+
+  it("After a successful submission, the form disappears and a screen with `You created a new box` appears with a box-id", async () => {
+    const submitBtn = component.getByRole("button", { name: /do the mutation/i });
+
+    fireEvent.click(submitBtn);
+
+    await waitFor(() => {
+      expect(component.queryByTestId("createBoxForm")).toBeNull();
+      expect(component.getByTestId("createdBox")).toBeTruthy();
+      expect(component.getByRole("heading", { name: /the box id is: 456/i })).toHaveTextContent(
```
Comment (HaGuesto) [inline react/src/views/CreateBox/CreateBox.test.js:107]: this part feels weird and looks like duplication
```
{ name: /the box id is: 456/i }
```

File: react/src/views/CreateBox/CreateBox.test.js [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+describe("Network error after submission", () => {
+  let component = null;
+  beforeEach(() => {
+    const history = createMemoryHistory();
+    const state = { qr: "barcode=387b0f0f5e62cebcafd48383035a92a" };
+    history.push("/create-box", state);
+
+    component = render(
+      <MockedProvider
+        mocks={[
```
Comment (HaGuesto) [inline react/src/views/CreateBox/CreateBox.test.js:171]: this should be defined in the beginning of the doc and passed into a utility function

File: react/src/views/CreateBox/CreateBox.test.js [hunk 6]
```diff
@@ -0,0 +1,270 @@
...
+describe("GraphQL error after submission", () => {
+  let component = null;
+  beforeEach(() => {
+    const history = createMemoryHistory();
+    const state = { qr: "barcode=387b0f0f5e62cebcafd48383035a92a" };
+    history.push("/create-box", state);
+
+    component = render(
+      <MockedProvider
+        mocks={[
```
Comment (HaGuesto) [inline react/src/views/CreateBox/CreateBox.test.js:218]: same here

File: react/src/views/CreateBox/CreateBox.test.js [hunk 7]
```diff
@@ -0,0 +1,270 @@
...
+
+    fireEvent.click(submitBtn);
+
+    await waitFor(() => {
+      expect(component.getByText("Error :( Please try again")).toBeInTheDocument();
+    });
+  });
+});
+
+// Loading state is a work in progress
```
Comment (HaGuesto) [inline react/src/views/CreateBox/CreateBox.test.js:258]: please add a comment why it did not work

File: react/src/views/Organization/SingleBase.tsx [hunk 8]
```diff
@@ -30,8 +30,8 @@ export default function SingleBase() {
     <div className="p-6">
```
Comment (HaGuesto) [inline react/src/views/Organization/SingleBase.tsx:30]: this file should not be included in the PR

**Review by HaGuesto — CHANGES_REQUESTED** [review]
Hey @billyhao12, thank you for the PR. Sry for the delay in reviewing!

*Functional Review*
whoop whoop! It runs through and should do what it is supposed to. The empty form fields test should be changed in the future.

*Code Review*
I left a few things to improve. :)

File: react/package.json [hunk 9]
```diff
@@ -70,7 +70,9 @@
     "flow-typed": "^3.1.0",
     "husky": "^4.2.5",
     "lint-staged": "^10.2.2",
+    "mutationobserver-shim": "^0.3.7",
```
[the remark this answers is not in the document]
Comment (billyhao12) [reply react/package.json:73]: Yes, all the tests fail without it.

File: react/src/views/CreateBox/CreateBox.test.js [hunk 10]
```diff
@@ -0,0 +1,270 @@
+import React from "react";
+import { render, waitFor, fireEvent, cleanup } from "@testing-library/react";
+import "@testing-library/jest-dom";
+import { MockedProvider } from "@apollo/client/testing";
+import CreateBox from "./CreateBox";
+import { CREATE_BOX } from "../../utils/queries";
+import { Router } from "react-router-dom";
+import { createMemoryHistory } from "history";
+import "mutationobserver-shim";
```
[the remark this answers is not in the document]
Comment (billyhao12) [reply react/src/views/CreateBox/CreateBox.test.js:9]: Yes, all of these imports are being used in the tests.

File: react/src/views/CreateBox/CreateBox.test.js (hunk 4 above)
Comment (billyhao12) [reply react/src/views/CreateBox/CreateBox.test.js:107]: Would it be more succinct to do this?
```
expect(component.findByRole("heading", { name: /the box id is: 456/i }));
```
This version would omit `.toHaveTextContent()`.
The findBy query returns a resolved Promise when an element is found: https://testing-library.com/docs/queries/about/#types-of-queries

File: react/package.json [hunk 11]
```diff
@@ -13,7 +13,7 @@
     "@progress/kendo-drawing": "^1.7.0",
     "@progress/kendo-react-pdf": "^3.14.0",
     "@testing-library/jest-dom": "^4.2.4",
-    "@testing-library/react": "^9.3.2",
+    "@testing-library/react": "11.2.2",
```
[the remark this answers is not in the document]
Comment (billyhao12) [reply react/package.json:16]: Should I do the same for @testing-library/jest-dom and @testing-library/user-event?

File: react/src/views/CreateBox/CreateBox.test.js (hunk 7 above)
Comment (billyhao12) [reply react/src/views/CreateBox/CreateBox.test.js:258]: Is it enough if I include this link?: https://github.com/apollographql/apollo-client/issues/6383
If not, I can provide more details.

File: react/package.json (hunk 11 above)
[the remark this answers is not in the document]
Comment (HaGuesto) [reply react/package.json:16]: yes

File: react/src/views/CreateBox/CreateBox.test.js (hunk 4 above)
Comment (HaGuesto) [reply react/src/views/CreateBox/CreateBox.test.js:107]: yes why not. not a biggy though

**Comment by HaGuesto** [comment]
@billyhao12 I believe you deleted the files I wanted you to take out from the PR. 

u should probably revert the last commit. If u are not 100% certain with git how to remove the changes to the files I marked from the PR, then I am also fine with it.
These changes are just formatting and we will do an overall re-formatting of all files soon.

**Review by HaGuesto — APPROVED** [review]
looks good now and still works. Thanks @billyhao12
